### PR TITLE
Pendual: Finish the majority of the events

### PR DESCRIPTION
**Pull request by trmazi**
In this commit I have pushed my current changes to IIDX 22 Pendual. I've completed every event other than Qpronicle Chord as I have ran out of steam on this mix for now. I will return and try to 100% this game sooner than later though. For now i'll leave it at this.

I have made the events all Frontend modifiable so that's good enough for now. I've thrown in a ton of #TODO statements to show my current progress.

### CODE REVIEW

**Comment by DragonMinded** [comment]
This is kind-of a mess but I'll try to review it anyway. General comments: why are you submitting this with a bunch of TODOs added when it is adding events for a single mix? Surely you could button it up before sending it in. As it stands its not mergeable.

File: bemani/backend/iidx/pendual.py [hunk 1]
```diff
@@ -134,10 +134,10 @@ def get_settings(cls) -> Dict[str, Any]:
                     'setting': 'global_shop_ranking',
                 },
                 {
-                    'name': 'Events In Omnimix',
```
Comment (DragonMinded) [inline bemani/backend/iidx/pendual.py:137]: Why did you get rid of this switch? Every other IIDX supports disabling events in omnimix.

File: bemani/backend/iidx/pendual.py [hunk 2]
```diff
@@ -153,6 +153,18 @@ def get_settings(cls) -> Dict[str, Any]:
...
                 },
+                {
+                    'name': 'Event Phase',
+                    'tip': 'Sets the machine event phase.',
+                    'category': 'game_config',
+                    'setting': 'event_phase',
+                    'values': {
+                        0: 'No Event',
+                        1: 'Chrono Seeker',
+                        #2: 'Qpronicle Chord', TODO: Finish this event
```
Comment (DragonMinded) [inline bemani/backend/iidx/pendual.py:170]: What's remaining?

File: bemani/backend/iidx/pendual.py [hunk 3]
```diff
@@ -889,46 +900,40 @@ def handle_IIDX22pc_request(self, request: Node) -> Optional[Node]:
 
             limit = Node.void('limit')
             root.add_child(limit)
-            limit.set_attribute('phase', '24')
+            limit.set_attribute('phase', '21')
```
Comment (DragonMinded) [inline bemani/backend/iidx/pendual.py:909]: Is this the real max?

File: bemani/backend/iidx/pendual.py [hunk 4]
```diff
@@ -889,46 +900,40 @@ def handle_IIDX22pc_request(self, request: Node) -> Optional[Node]:
...
                 game_config = self.get_game_config()
                 timeshift_override = game_config.get_int('cycle_config')
-                omni_events = game_config.get_bool('omnimix_events_enabled')
+                ccxbm_on = game_config.get_bool('ccxbm_enable')
             else:
                 # If we aren't in an arcade, we turn off events
                 timeshift_override = 0
-                omni_events = False
-
-            if self.omnimix and (not omni_events):
```
Comment (DragonMinded) [inline bemani/backend/iidx/pendual.py:904]: Second comment on this, why remove this?

File: bemani/backend/iidx/pendual.py [hunk 5]
```diff
@@ -889,46 +900,40 @@ def handle_IIDX22pc_request(self, request: Node) -> Optional[Node]:
...
 
             qpronicle_chord = Node.void('qpronicle_chord')
             root.add_child(qpronicle_chord)
-            qpronicle_chord.set_attribute('phase', '2')
+            qpronicle_chord.set_attribute('phase', '3')
 
             common_cd_event = Node.void('common_cd_event')
             root.add_child(common_cd_event)
-            common_cd_event.set_attribute('open_list', '3')
+            common_cd_event.set_attribute('open_list', str(3))
```
Comment (DragonMinded) [inline bemani/backend/iidx/pendual.py]: Why this change?

File: bemani/backend/iidx/pendual.py [hunk 6]
```diff
@@ -1185,6 +1190,35 @@ def handle_IIDX22pc_request(self, request: Node) -> Optional[Node]:
...
+            root.add_child(tohoisfumo)
+            tohoisfumo.set_attribute('phase', "2")
+
+            cm = Node.void("cm")
+            root.add_child(cm)
+            cm.set_attribute('compo', 'cm_event3')
+            cm.set_attribute('folder', 'cm_event3')
+            cm.set_attribute('id', "0")
+
+            #Ranking details in title screen
```
Comment (DragonMinded) [inline bemani/backend/iidx/pendual.py:1229]: Nice.

File: bemani/backend/iidx/pendual.py [hunk 7]
```diff
@@ -1185,6 +1190,35 @@ def handle_IIDX22pc_request(self, request: Node) -> Optional[Node]:
...
+            rankingcharts = []
+            for (mid, plays) in self.data.local.music.get_hit_chart(self.game, self.music_version, 20):
+                rankingcharts.append(mid)
+            root.add_child(Node.u16_array('monthly_mranking', rankingcharts))
+            root.add_child(Node.u16_array('total_mranking', rankingcharts))
+            ##need to add in the top 20 rankings for month and total network
```
Comment (DragonMinded) [inline bemani/backend/iidx/pendual.py:1235]: Is that not what the above section does? Except for it doesn't really do it for the last month.

File: bemani/backend/iidx/pendual.py [hunk 8]
```diff
@@ -1707,6 +1741,172 @@ def format_profile(self, userid: UserID, profile: Profile) -> Node:
         # Ea app features
         if self.data.triggers.has_broadcast_destination(self.game):
             root.add_child(Node.void('bind_eaappli'))
+
+        #cokeXBEMANI event details
+        cxb_music = Node.void('cc_collabo_music')
+        root.add_child(cxb_music)
+        cxb_music.set_attribute('music_list', "1")
+        #TODO: Find the proper music_list number
```
Comment (DragonMinded) [inline bemani/backend/iidx/pendual.py]: Seems this isn't finished? Did you check bemaniwiki?

File: bemani/backend/iidx/pendual.py [hunk 9]
```diff
@@ -1707,6 +1741,172 @@ def format_profile(self, userid: UserID, profile: Profile) -> Node:
...
+        #SOUND HOLIC Event
+        soundhoic = Node.void('sound_holic')
+        root.add_child(soundhoic)
+        soundhoic.set_attribute('music_list', "1")
+        soundhoic.set_attribute('announce_list', "1")
+
+        #sandglass event details
+        time_sandglass = Node.void('time_sandglass')
+        root.add_child(time_sandglass)
+        time_sandglass.set_attribute('item_num', "99")
```
Comment (DragonMinded) [inline bemani/backend/iidx/pendual.py:1788]: So this isn't done at all? Is it going to be done like infinite paseli checks?

File: bemani/backend/iidx/pendual.py [hunk 10]
```diff
@@ -1707,6 +1741,172 @@ def format_profile(self, userid: UserID, profile: Profile) -> Node:
...
+
+        #konami pyramid scheme
+        pyramid = Node.void('pyramid')
+        root.add_child(pyramid)
+        pyramid.set_attribute('music_list', "1")
+        pyramid.set_attribute('item_list', "1")
+        pyramid.set_attribute('statue_0', str(0))
+        pyramid.set_attribute('statue_1', str(0))
+        pyramid.set_attribute('statue_2', str(0))
+        #TODO: Figure out how to actually send this data.
```
Comment (DragonMinded) [inline bemani/backend/iidx/pendual.py]: I gave up reviewing here. This seems like a bunch of hardcoded stuff to make the event *show up* but it doesn't actually implement any of the logic to make the event *work*.

**Review by DragonMinded — CHANGES_REQUESTED** [review]
I don't think you ran lint or type checking? There's stuff I can see at a glance would set that off.

File: bemani/backend/iidx/pendual.py (hunk 1 above)
Comment (trmazi) [reply bemani/backend/iidx/pendual.py:137]: I removed it and had forgotten to re-implement it. Will be fixed in the next commit.

File: bemani/backend/iidx/pendual.py (hunk 2 above)
Comment (trmazi) [reply bemani/backend/iidx/pendual.py:170]: I need to figure out the achievements section of the event. For some reason the game kills off the event if I send through it after saving in unformat_profile.

File: bemani/backend/iidx/pendual.py (hunk 3 above)
Comment (trmazi) [reply bemani/backend/iidx/pendual.py:909]: from my understanding, yes.

File: bemani/backend/iidx/pendual.py (hunk 4 above)
Comment (trmazi) [reply bemani/backend/iidx/pendual.py:904]: See explanation above in the first comment.

File: bemani/backend/iidx/pendual.py (hunk 5 above)
Comment (trmazi) [reply bemani/backend/iidx/pendual.py]: I had messed around with changing the values with frontend, meant to set it back to just being a string.

File: bemani/backend/iidx/pendual.py (hunk 6 above)
Comment (trmazi) [reply bemani/backend/iidx/pendual.py:1229]: these are separate: comment goes to the rankings under, though this is not actually what they're for or im missing part of the puzzle. was meant to be looked at again, had not gotten to that.

File: bemani/backend/iidx/pendual.py (hunk 7 above)
Comment (trmazi) [reply bemani/backend/iidx/pendual.py:1235]: need to fix this with the proper values

File: bemani/backend/iidx/pendual.py (hunk 8 above)
Comment (trmazi) [reply bemani/backend/iidx/pendual.py]: I did not think that there would be any information there about the coke event music details. I plan on finishing that properly in the next commit.

File: bemani/backend/iidx/pendual.py (hunk 9 above)
Comment (trmazi) [reply bemani/backend/iidx/pendual.py:1788]: Yes, this is set to infinite. Konami added sandglasses to your profile by buying them with paseli and the game removes them from your account at logout. I could make these purchasable on frontend similar to paseli, but for now they're set to infinite.

File: bemani/backend/iidx/pendual.py (hunk 10 above)
Comment (trmazi) [reply bemani/backend/iidx/pendual.py]: This does not cause the pyramid event to show up. i suppose i should have removed them from my PR as they're basically useless as they stand and are not required. From what i can tell, this event will only trigger on an older build of pendual, so i might as well just not bother with it at all.

**Comment by Subject38** [comment]
Having looked over at the changes, everything looks fine to me, honestly confused why there hasn't been further review...

**Review by DragonMinded — APPROVED** [review]
